solr/core/src/java/org/apache/solr/search/combining/CombineQParserPlugin.java

```diff
@@ -0,0 +1,167 @@
+
+      for (String queryKey : queriesToCombineKeys) {
+        final var unparsedQuery = blendParams.get(queryKey);
+        final var parser = QParser.getParser(unparsedQuery, defType, req);
+        var query = parser.getQuery();
+        if (query == null) { // sad this can happen
+          query = new MatchNoDocsQuery();
+        }
+        unparsedQueries.add(unparsedQuery);
+        queryParsers.add(parser);
+        queries.add(query);
+      }
+
+      return new CombineQuery(queriesCombiningStrategy, queries);
+    }
+
+    @Override
+    public void addDebugInfo(NamedList<Object> dbg) {
+      super.addDebugInfo(dbg);
+
+      NamedList<NamedList<Object>> queriesDebug = new SimpleOrderedMap<>();
+      String[] queryKeys = req.getParams().getParams(CombinerParams.COMBINER_KEYS);
+      for (int i = 0; i < queries.size(); i++) {
+        NamedList<Object> singleQueryDebug = new SimpleOrderedMap<>();
+        singleQueryDebug.add("querystring", unparsedQueries.get(i));
+        singleQueryDebug.add("queryparser", queryParsers.get(i).getClass().getSimpleName());
+        singleQueryDebug.add("parsedquery", QueryParsing.toString(queries.get(i), req.getSchema()));
+        singleQueryDebug.add("parsedquery_toString", queries.get(i).toString());
+        queriesDebug.add(queryKeys[i], singleQueryDebug);
+      }
+      dbg.add("queriesToCombine", queriesDebug);
+    }
+  }
+
+  static final class CombineQuery extends ExtendedQueryBase
+      implements QueryComponent.SelfExecutingQuery {
+
+    private final QueriesCombiner combiner;
+    private final List<Query> queries;
+
+    public CombineQuery(QueriesCombiner combiner, List<Query> queries) {
+      this.combiner = combiner;
+      this.queries = queries;
+    }
+
+    @Override
+    public String toString(String field) {
+      return super.toString(field) + "{!" + NAME + "}"; // TODO others
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) return true;
+      if (!(o instanceof CombineQuery)) return false;
+      CombineQuery that = (CombineQuery) o;
+      return Objects.equals(combiner, that.combiner) && Objects.equals(queries, that.queries);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(combiner, queries);
+    }
+
+    @Override
+    public void visit(QueryVisitor visitor) {
+      for (Query query : queries) {
+        query.visit(visitor.getSubVisitor(Occur.MUST, this));
+      }
+    }
+
+    @Override
+    public QueryResult search(SolrIndexSearcher searcher, QueryCommand cmd) throws IOException {
+      QueryResult[] results = new QueryResult[queries.size()];
+      // TODO do in multiple threads?
+      for (int i = 0; i < queries.size(); i++) {
+        cmd.setQuery(queries.get(i));
+        QueryResult qr = new QueryResult();
+        searcher.search(qr, cmd);
+        results[i] = qr;
+      }
+
+      // nocommit but how is the docSet (e.g. for faceting) or maybe other things supported?
```

Comment:
This seems like a fundamental limitation in the whole feature honestly.  What Hossman said about re-rank query is interesting... not sure if it addresses this.

Reply:
mmm probably I have lost in the comments why the way I created the docSet was incorrect. Can you elaborate? (a nice occasion to learn how that docSet part works, as I didn't have time yet)